src/main/scala/org/constellation/domain/transaction/TransactionChainService.scala

```diff
@@ -0,0 +1,44 @@
+package org.constellation.domain.transaction
+
+import cats.effect.Concurrent
+import cats.implicits._
+import io.chrisdavenport.log4cats.Logger
+import org.constellation.primitives.concurrency.SingleRef
+
+class TransactionChainService[F[_]: Concurrent: Logger] {
+  // TODO: Make sure to clean-up those properly
+  private[domain] val lastTransactionCount: SingleRef[F, Long] = SingleRef(0)
+  private[domain] val addressCount: SingleRef[F, Map[String, Long]] = SingleRef(Map.empty[String, Long])
+  private[domain] val lastTxHash: SingleRef[F, Map[String, String]] = SingleRef(Map.empty[String, String])
+
+  def getNext(address: String): F[(String, Long)] =
+    lastTransactionCount
+      .modify(count => (count + 1, count + 1))
+      .flatMap(count => lastTransactionHash(address).map(_.getOrElse("")).map(a => (a, count)))
+
+  def getLatest(address: String): F[Long] =
+    addressCount.get
+      .map(_.getOrElse(address, 0))
+
+  def lastTransactionHash(address: String): F[Option[String]] =
```

Comment:
`getLastTransactionHash` -> `lastTransactionHash`

Reply:
there is `lastTransactionHash`